src/frequenz/sdk/microgrid/metadata.py

```diff
@@ -0,0 +1,54 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Metadata that describes a microgrid."""
+
+from dataclasses import dataclass
+from zoneinfo import ZoneInfo
+
+from timezonefinder import TimezoneFinder
+
+_timezone_finder = TimezoneFinder()
+
+
+@dataclass
+class Location:
+    """Metadata for the location of microgrid."""
+
+    latitude: float | None = None
+    """The latitude of the microgrid in degree."""
+
+    longitude: float | None = None
+    """The longitude of the microgrid in degree."""
+
+    _timezone: ZoneInfo | None = None
+    """The timezone of the microgrid.
+
+    The timezone will be set to None if the latitude or longitude points
+    are not set or the timezone cannot be found given the location points.
+    """
+
+    def __post_init__(self) -> None:
+        """Initialize the timezone of the microgrid."""
+        if self.latitude is None or self.longitude is None:
+            return
+
+        timezone = _timezone_finder.timezone_at(lat=self.latitude, lng=self.longitude)
+        if timezone:
+            self._timezone = ZoneInfo(key=timezone)
```

Comment:
Ah, I guess no frozen because you need to modify it here, but there is a hack:

```suggestion
    timezone: ZoneInfo | None = None
    """The timezone of the microgrid.

    The timezone will be set to None if the latitude or longitude points
    are not set or the timezone cannot be found given the location points.
    """

    def __post_init__(self) -> None:
        """Initialize the timezone of the microgrid."""
        if self.latitude is None or self.longitude is None or self.timezone is not None:
            return

        timezone = _timezone_finder.timezone_at(lat=self.latitude, lng=self.longitude)
        if timezone:
            object.__setattr__(self, "timezone", ZoneInfo(key=timezone))
```

With this also the user can manually set the timezone too (suppose that in the future the microgrid API provides it directly). This could be also useful for testing.

Reply:
yeah the reason was setting `self._timezone`  at `__post_init__`, though I was not aware of such a hack python:registered:

Reply:
In some sense Python is like C, you can always access the underlying pointers and mess with the raw data :D